Add Microsoft Hypervisor as a supported hypervisor

russell-islam opened the pull request.

liuw commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -0,0 +1,10 @@
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+#![allow(unused_macros)]
+#![allow(non_upper_case_globals)]
```
These were added to help focus on development. At the time I implemented more things than necessary because I wanted them to be available whenever I needed them.

At least the first four should be dropped when it comes to upstreaming and we should fix whatever warnings that show up. I don't think there are that many.

russell-islam replied:
I was thinking to do it in a future PR if you do agree.

liuw replied:
Fine by me.

liuw commented on hypervisor/src/mshv/x86_64/mod.rs:
```diff
@@ -0,0 +1,154 @@
+        msr!(msr_index::MSR_KERNEL_GS_BASE),
+        msr!(msr_index::MSR_SYSCALL_MASK),
+        msr!(msr_index::MSR_IA32_TSC),
+        // TODO: feature guard kvm and move this out of kvm
```
This comment seems to be stale.

rbradford replied:
Did you miss acting on this feedback?

liuw commented on hypervisor/src/lib.rs:
```diff
@@ -62,5 +62,5 @@ pub fn new() -> std::result::Result<Arc<dyn Hypervisor>, HypervisorError> {
     Ok(Arc::new(hv))
 }
-#[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+
```
Unrelated change to this particular patch.

(Not sure how this will show up in GitHub UI)

liuw replied:
Also, there is an extraneous blank line between the two SoB lines.

liuw commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -13,6 +13,18 @@
+pub use mshv_bindings::*;
+use mshv_ioctls::{set_registers_64, Mshv, VcpuFd, VmFd};
+
+/// Vcpu struct for Hyper-V
```
Hyper-V should be Microsoft Hypervisor. Please also fix other occurrences in code and commit message.

liuw replied:
And again, there is an extraneous line between two SoB lines.

liuw commented on hypervisor/src/vm.rs:
```diff
@@ -220,6 +229,18 @@ pub trait Vm: Send + Sync {
     fn check_extension(&self, c: Cap) -> bool;
     /// Create a device that is used for passthrough
     fn create_passthrough_device(&self) -> Result<Arc<dyn Device>>;
+    /// XXX temp workaround for Hyper-V, never used on KVM
```
Please turn this `XXX` into a proper comment.

liuw commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -586,6 +586,117 @@ impl cpu::Vcpu for MshvVcpu {
     }
 }
 
+struct MshvEmulatorContext<'a> {
```
Sameul and I wrote this patch so you should drop Nuno and Praveen's tags.

liuw reviewed (commented):
Thanks for doing this, Muminul.

I don't have much to say since I've written a big portion of this series, so of course I approved the general direction. ;-)

I commented on individual patches. Not sure how they will show up in GitHub's UI, but if you can, navigate to individual patches may make my comments make more sense to you.

sameo commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
+fn get_delivery_mode(message_data: u32) -> u8 {
+    ((message_data & 0x700) >> 8) as u8
+}
+
```
Maybe it would be cleaner to make those `MshvIrqRoutingMsi` methods instead of having them floating around the code?

russell-islam replied:
This implementation mainly contributed by @liuw >I would let him comment.

liuw replied:
Yes. Making them methods of `MshvIrqRoutingMsi` is fine. They are not tied to MSHV in anyway, but making them methods won't hurt either.

sameo commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
+
+    false
+}
+
```
I think this one is not used.

sameo commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -526,7 +526,154 @@ impl cpu::Vcpu for MshvVcpu {
+                        let v = data[0] as u32
+                            | (data[1] as u32) << 8
+                            | (data[2] as u32) << 16
+                            | (data[3] as u32) << 24;
```
This is `let v = u32::from_le_bytes(data);`

sameo commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -526,7 +526,154 @@ impl cpu::Vcpu for MshvVcpu {
+                        data[0] = info.rax as u8;
+                        data[1] = (info.rax >> 8) as u8;
+                        data[2] = (info.rax >> 16) as u8;
+                        data[3] = (info.rax >> 24) as u8;
```
`let data = (info.rax as u32).to_le_bytes()`

russell-islam commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
```
Not really, we followed the same as before. If you think this is something we need to revisit please sync up with Guru.

russell-islam replied:
@gbalane Can also comment here.

russell-islam replied:
We can revisit this later if needed.

gbalane replied:
> @gbalane Can also comment here.

We are fine with Dual Licensing and it is inline with CLH based on the previous discussion.

russell-islam commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
+                "IrqfdCtrlEpollHandler: failed to register listener: {:?}",
+                err
+            );
+        });
```
I would let @liuw  to comment.

liuw replied:
No. I don't think it makes `run_ctrl()` exit. The thread will be just hanging forever because it will block on those eventfds.

I was not sure at the time there would be a better way to handle this. I didn't want to exit the thread because I was worried that it might cause failure when `unregister_irqfd` wrote to the `kill_fd`. Now that I think about this, it is probably okay to let the thread exit because `kill_fd` will hang on for as long as needed.

Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

sameo replied:
> Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

Then let's keep it as is for now.

liuw commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
+
+// See Intel SDM vol3 10.11.1
+// We assume APIC ID and Hyper-V Vcpu ID are the same value
+// This holds true for HvLite
```
HvLite is a project that is internal to Microsoft. This comment should be dropped.

russell-islam commented on hypervisor/src/kvm/mod.rs:
```diff
@@ -159,7 +159,7 @@ impl vm::Vm for KvmVm {
     /// Registers an event that will, when signaled, trigger the `gsi` IRQ.
     ///
-    fn register_irqfd(&self, fd: &EventFd, gsi: u32) -> vm::Result<()> {
+    fn register_irqfd(&self, fd: &EventFd, gsi: u32, _vm: Arc<dyn vm::Vm>) -> vm::Result<()> {
```
I do agree with you @sameo. Let get thought from @liuw

liuw replied:
I agree with Samuel. The code could use a bit of restructuring.

liuw commented:
@sameo @rbradford 

The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.

I've written a patch. See https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de .

I would like some input from you both on whether that's acceptable.

sameo commented:
> @sameo @rbradford
> 
> The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.
> 
> I've written a patch. See [liuw@6df3afd](https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de) .
> 
> I would like some input from you both on whether that's acceptable.

What about that one https://github.com/sameo/cloud-hypervisor-1/commit/3cd4916ce1283d7bdd410027d2ba1865b199c4b6 ?

liuw commented:
> 
> 
> > @sameo @rbradford
> > The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.
> > I've written a patch. See [liuw@6df3afd](https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de) .
> > I would like some input from you both on whether that's acceptable.
> 
> What about that one [sameo@3cd4916](https://github.com/sameo/cloud-hypervisor-1/commit/3cd4916ce1283d7bdd410027d2ba1865b199c4b6) ?

This looks better. Let's go with yours.

russell-islam commented:
@sameo @liuw The clippy warnigs have been resolved now. Only error we are geeting is as follow 
`thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_middle/src/ty/query/mod.rs:235:5`
I this this is a known issue upstream.

Also I have addressed all of your comments. If I missed anything that is not intentional. Please let me know if you still have some comments. I will address those ASAP and try to get this PR merged at my earliest.

liuw commented on hypervisor/src/cpu.rs:
```diff
@@ -272,10 +272,12 @@ pub trait Vcpu: Send + Sync {
     /// Setup the model-specific registers (MSR) for this vCPU.
     ///
     fn set_msrs(&self, msrs: &MsrEntries) -> Result<usize>;
+    #[cfg(feature = "kvm")]
```
There is a word "bas" which should perhaps be deleted?

russell-islam commented on hypervisor/src/mshv/x86_64/mod.rs:
```diff
@@ -0,0 +1,154 @@
+    /// Representation of an memory mapped I/O address.
+    Mmio(u64),
+}
+macro_rules! msr {
```
Will do in a future PR.

russell-islam commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -464,12 +478,13 @@ impl vm::Vm for MshvVm {
     ///
     fn state(&self) -> vm::Result<VmState> {
-        unimplemented!();
+        Ok(*self.hv_state.read().unwrap())
```
Yes. This is needed for restore and snapshot. This is VM specific data.  hv_state is named after the hv interface documentationjs.

sameo commented:
We worked with @russell-islam on the following, additional changes:

- Clean the commit messages up in order for them to adhere to https://www.kernel.org/doc/Documentation/process/submitting-patches.rst wrt to SOBs and Co-developed-by tags
- Squash 35bf055e27f7867654a506fa0c46302646d434a2 with 4bb31daa18e5d7181c7d4e1388e101b5536db463
- Include missing Cargo.lock changes

russell-islam commented:
> We worked with @russell-islam on the following, additional changes:
> 
>     * Clean the commit messages up in order for them to adhere to https://www.kernel.org/doc/Documentation/process/submitting-patches.rst wrt to SOBs and Co-developed-by tags
> 
>     * Squash [35bf055](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/35bf055e27f7867654a506fa0c46302646d434a2) with [4bb31da](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/4bb31daa18e5d7181c7d4e1388e101b5536db463)
> 
>     * Include missing Cargo.lock changes

Thanks @sameo  and @liuw  for helping all the way. The PR looks good to me. I have listed all the future TODO that I am planning to finish gradually.

rbradford commented:
Can you open a GitHub issue with all the things that have been postponed to a later PR?

rbradford commented on hypervisor/src/mshv/x86_64/mod.rs:
```diff
@@ -0,0 +1,154 @@
+            msr_index::MSR_IA32_MISC_ENABLE_FAST_STRING as u64
+        ),
+        msr_data!(msr_index::MSR_MTRRdefType, MTRR_ENABLE | MTRR_MEM_TYPE_WB), */
+    ])
```
Commented out code. Why is this here?

liuw replied:
This should've been dropped.

rbradford commented on hypervisor/src/mshv/x86_64/mod.rs:
```diff
@@ -0,0 +1,154 @@
+    mshv_bindings::StandardRegisters, mshv_bindings::VcpuEvents, mshv_bindings::XSave as Xsave,
+    mshv_bindings::Xcrs as ExtendedControlRegisters,
+};
+#[derive(Clone, Serialize, Deserialize)]
```
I think there should be a line of whitespace between the use/pub use and also between the pub use and the struct definitions.

rbradford commented on hypervisor/src/cpu.rs:
```diff
@@ -12,13 +12,18 @@
+use crate::CpuState;
+#[cfg(target_arch = "x86_64")]
+use crate::Xsave;
 
```
This whitespace line breaks the ordering.

rbradford commented on hypervisor/Cargo.toml:
```diff
@@ -26,6 +26,8 @@ serde_json = ">=1.0.9"
 vm-memory = { version = "0.4.0", features = ["backend-mmap", "backend-atomic"] }
 vmm-sys-util = { version = ">=0.5.0", features = ["with-serde"] }
 
+epoll = ">=4.0.1"
```
This should be in alphabetical order without the whitespace before it

rbradford commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -17,7 +17,7 @@ use crate::cpu::Vcpu;
 pub use mshv_bindings::*;
-use mshv_ioctls::{set_registers_64, Mshv, VcpuFd, VmFd};
+use mshv_ioctls::{set_registers_64, InterruptRequest, Mshv, VcpuFd, VmFd};
 
```
Whitespace breaks the rustfmt lexical ordering

rbradford commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -35,7 +35,6 @@ pub use x86_64::*;
 
```
Ditto.

rbradford commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -35,7 +35,6 @@ pub use x86_64::*;
 
 use crate::device;
 
```
Ditto

liuw reviewed (approved):
This series LGTM.

Rob made a few cosmetic comments and I also found a typo in a commit message. To save another round-trip I've addressed Rob's comments and fixed the typo, and pushed to Muminul's branch.

liuw commented:
@russell-islam Please create an issue to track future PRs like Rob suggested.

russell-islam commented:
Created an issue to track followup PRs. https://github.com/cloud-hypervisor/cloud-hypervisor/issues/2048
